libpromises/evalfunction.c

```diff
@@ -1188,6 +1188,95 @@ static FnCallResult FnCallGetGid(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const P
 #endif
 }
 
+/*********************************************************************/ 
+
+static FnCallResult FnCallUserInGroup(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
+{
+    assert(fp != NULL);
+#ifdef _WIN32
+    Log(LOG_LEVEL_ERR, "Function '%s' is POSIX specific", fp->name);
+    return FnFailure();  
+#else
+    assert(finalargs != NULL);
+
+    const char *user_name = RlistScalarValue(finalargs);
+    const char *group_name = RlistScalarValue(finalargs->next);
+
+    StringSet *users = StringSetNew();
+    int ret;
+
+    // check secondary group
+    struct group grp;
+    struct group *grent;
+    char gr_buf[GETGR_R_SIZE_MAX] = {0};
+    ret = getgrnam_r(group_name, &grp, gr_buf, GETGR_R_SIZE_MAX, &grent);
+    
+    if (grent == NULL)
+    {
+        switch(ret)
+        {
+            case 0:
+            case ENOENT:
+            case ESRCH:
+            case EBADF:
+            case EPERM:
+                break;
+            default:
+            {
+                const char *error_msg = GetErrorStrFromCode(ret);
+                Log(LOG_LEVEL_ERR, "Couldn't open group database in function '%s': %s", fp->name, error_msg);
+                StringSetDestroy(users);
+                return FnFailure();
+            }
+        }
+    }
+    else {
+        while (grent->gr_mem[0] != NULL)
+        {
+            char *group_member = xstrdup(grent->gr_mem[0]);
+            StringSetAdd(users, group_member);
+            grent->gr_mem++;
```

Comment:
Here it would be more optimized to just check if the group member is equal to the user we are looking for. If it is we can return early. There is no need to build a string set, and then iterate over it again for `StringSetContains()`. Also there is no need to iterate over the rest of the members after we have found our match. Just return early here if we find the user. Remember to clean up / free before returning.